core/src/main/java/com/linecorp/armeria/internal/common/stream/AppendingPublisher.java

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.common.stream;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
+
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import com.google.common.math.LongMath;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.stream.NoopSubscriber;
+
+public final class AppendingPublisher<T> implements Publisher<T> {
+
+    private final Publisher<? extends T> publisher;
+    private final T append;
+
+    public AppendingPublisher(Publisher<? extends T> publisher, T append) {
+        this.publisher = publisher;
+        this.append = append;
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super T> subscriber) {
+        requireNonNull(subscriber, "subscriber");
+        subscriber.onSubscribe(new AppendSubscriber<>(publisher, append, subscriber));
+    }
+
+    static final class AppendSubscriber<T> implements Subscriber<T>, Subscription {
+
+        @SuppressWarnings("rawtypes")
+        private static final AtomicLongFieldUpdater<AppendSubscriber> requestedUpdater =
+                AtomicLongFieldUpdater.newUpdater(AppendSubscriber.class, "requested");
+
+        @SuppressWarnings("rawtypes")
+        private static final AtomicLongFieldUpdater<AppendSubscriber> needToPublishUpdater =
+                AtomicLongFieldUpdater.newUpdater(AppendSubscriber.class, "needToPublish");
+
+        private final Publisher<? extends T> publisher;
+        private final T append;
+        private Subscriber<? super T> downstream;
+        @Nullable
+        private volatile Subscription upstream;
+
+        private volatile long requested;
+        private volatile long needToPublish;
+        private boolean subscribed;
+        private volatile boolean cancelled;
+
+        private boolean appendNeedToSend;
+        private boolean appendSent;
+        private boolean completeSent;
```

Comment:
To pass `TCK` test, we need theses variables but seems too complicate..
If you have a good idea, feel free to share to me! 🙇